### PR TITLE
vo_dmabuf_wayland: use single-pixel-buffer-v1

### PR DESCRIPTION
**Pull request by LaserEyess**
Supercedes #10265

### CODE REVIEW

File: video/out/vo_dmabuf_wayland.c [hunk 1]
```diff
@@ -41,6 +41,9 @@
 #include "present_sync.h"
 #include "wayland_common.h"
 #include "generated/wayland/linux-dmabuf-unstable-v1.h"
+#if HAVE_WAYLAND_PROTOCOLS_1_27
+#include "generated/wayland/single-pixel-buffer-v1.h"
+#endif
 #include "generated/wayland/viewporter.h"
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:47]: This probably could be organized a bit better. Mind moving all the generated headers down in a separate section with the `#if` in its own section below. Like how `wayland_common.c` does it basically.

File: video/out/vo_dmabuf_wayland.c [hunk 2]
```diff
@@ -241,21 +244,33 @@ static int reconfig(struct vo *vo, struct mp_image_params *params)
...
-        int height = 1;
-        int stride = MP_ALIGN_UP(width * 4, 16);
-        int fd = vo_wayland_allocate_memfd(vo, stride);
-        if (fd < 0)
-            return VO_ERROR;
-        p->solid_buffer_pool = wl_shm_create_pool(wl->shm, fd, height * stride);
-        if (!p->solid_buffer_pool)
-            return VO_ERROR;
-        p->solid_buffer = wl_shm_pool_create_buffer(p->solid_buffer_pool, 0, width, height, stride, WL_SHM_FORMAT_XRGB8888);
+    if (!p->solid_buffer) {
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:247]: Just carrying over comments from irc, but all of this stuff could all probably go in `preinit`.

File: video/out/wlbuf_pool.h [hunk 3]
```diff
@@ -16,6 +16,7 @@
  */
 
 struct wlbuf_pool_entry;
+struct zwp_linux_buffer_params_v1;
```
Comment (Dudemanguy) [inline video/out/wlbuf_pool.h:19]: Good catch, but I think I'd prefer to just include the headers here. This also has `struct vo_wayland_state` tucked away in it which works by luck. i.e.

```
#include "wayland_common.h"
#include "generated/wayland/linux-dmabuf-unstable-v1.h"
```

File: video/out/wayland_common.c [hunk 4]
```diff
@@ -1240,6 +1241,10 @@ static void registry_handle_add(void *data, struct wl_registry *reg, uint32_t id
     if (!strcmp(interface, wp_content_type_manager_v1_interface.name) && found++) {
         wl->content_type_manager = wl_registry_bind(reg, id, &wp_content_type_manager_v1_interface, 1);
     }
+
+    if (!strcmp(interface, wp_single_pixel_buffer_manager_v1_interface.name) && found++) {
+        wl->single_pixel_manager = wl_registry_bind(reg, id, &wp_single_pixel_buffer_manager_v1_interface, 1);
+    }
```
Comment (Dudemanguy) [inline video/out/wayland_common.c:1247]: Could you also add logging for when this isn't found in `preinit`? i.e. the `Compositor doesn't support the %s protocol!` stuff.

**Comment by Dudemanguy** [comment]
I can confirm this works fine in both cases (with `single-pixel-buffer` and without it).

File: video/out/vo_dmabuf_wayland.c [hunk 5]
```diff
@@ -354,6 +344,35 @@ static int preinit(struct vo *vo)
...
+#if HAVE_WAYLAND_PROTOCOLS_1_27
+        p->solid_buffer = wp_single_pixel_buffer_manager_v1_create_u32_rgba_buffer(
+            vo->wl->single_pixel_manager, 0, 0, 0, UINT32_MAX); /* R, G, B, A */
+#endif
+    } else {
+#if HAVE_WAYLAND_PROTOCOLS_1_27
+        MP_INFO(vo->wl,
+            "Compositor doesn't support the %s protocol, falling back to using wl-shm",
+            wp_single_pixel_buffer_manager_v1_interface.name);
+#endif
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:358]: Not exactly what I wanted. This check should be in `wayland_common.c` under where the `if (wl->content_type_manager) {` block is.

**Comment by Dudemanguy** [comment]
Also you need to destroy the `single_pixel_buffer_manager` in `vo_wayland_uninit`.

File: video/out/vo_dmabuf_wayland.c [hunk 6]
```diff
@@ -41,9 +41,17 @@
...
+// Generated from wayland-protocols
 #include "generated/wayland/linux-dmabuf-unstable-v1.h"
 #include "generated/wayland/viewporter.h"
 
+#if HAVE_WAYLAND_PROTOCOLS_1_27
+#include "generated/wayland/single-pixel-buffer-v1.h"
+#endif
+
+
+
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:54]: Nit, could you nuke these two newlines?

**Review by Dudemanguy — APPROVED** [review]
LGTM otherwise

**Comment by LaserEyess** [comment]
Tested with and without single-pixel on sway, it's working.

**Comment by Dudemanguy** [comment]
Thanks!

**Pull request merged**